[file_selector] Add Android support

gantelo opened the pull request:
This PR adds the Android implementation for the `file_selector` package.

Issue related: [[file_selector] Add Android support #110098](https://github.com/flutter/flutter/issues/110098)

We worked on the structure of the Android native code (using Java) and the Dart code to link the work to this native part.

We based our structure on the other available packages to avoid having too many discrepancies between them.

However, we split the native implementation into a file for each class to avoid being too cumbersome with the activity helpers and wrappers.

For the `dart` code, we extend the `FileSelectorPlatform` and override the main four methods:
 - openFile
 - openFiles
 - getDirectoryPath
 - getSavePath

Current implemented methods:
 - [x] openFile
 - [x] openFiles
 - [x] getDirectoryPath
 - [x] getSavePath

Each of these invokes the `MethodChannel('plugins.flutter.io/file_selector_android')` that is our native code.

Our native code has two main points:
 - FileSelectorPlugin: this class has the `LifeCycleObserver` and `ActivityState` classes to manage the plugin's activity and the result wrappers to respond on the platform thread.
 - FileSelectorDelegate: is the heavy lifting for the plugin.

We have implemented the overriding of `onMethodCall,` which does the following:

```
    switch (call.method) {
      case METHOD_GET_DIRECTORY_PATH:
        delegate.getDirectoryPath(call, result);
        break;
      case METHOD_GET_SAVE_PATH:
        delegate.getSavePath(call, result);
        break;
      case METHOD_OPEN_FILE:
        delegate.openFile(call, result);
        break;
      default:
        throw new IllegalArgumentException("Unknown method " + call.method);
    }
```

As for the tests, we added tests for both the dart and Java implementation of the methods.

![getDirectoryPath](https://user-images.githubusercontent.com/49167576/191831968-b232c9f3-25e8-402e-baf7-530b427e75cc.gif)
**getDirectoryPath naive functionality**


Please, any early feedback will be more than welcome to avoid rework in the future.

Note: we couldn't set the Confirm Button Text property because of an inability, at least in the action, to [setting other parameters than the extra initial uri](https://developer.android.com/reference/android/content/Intent#ACTION_OPEN_DOCUMENT_TREE). This means that, in Android, we'll have to stick with the button text being "USE THIS FOLDER" for now.

_As ACTION_OPEN_DOCUMENT_TREE is SDK 21+, this is our current support. We'll add support for 19+ once we have feedback and a well-defined structure for the remaining items._

> Note: we may need an `override: no versioning needed` since this is not the final version of the package, and it won't be yet deployed unless you say so.

> Note 2: we'll have to update the documentation to add the proper support for Android once we decide we can ship this.
> This is already being done in [PR #27](https://github.com/southworks/plugins/pull/27)

> Note 3: This PR contains the whole Android functionality; for a more detailed review method by method, we have some separate draft PRs against the main blob.
>
> You can check those here:
> [Open multiple files](https://github.com/southworks/plugins/pull/34)
> [Open a single file](https://github.com/southworks/plugins/pull/24)
> [Save a file](https://github.com/southworks/plugins/pull/23)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

gantelo committed:
Add file_selector support for Android

gantelo commented:
About the failing publishable check:
How should we manage this with a new package? We're adding support for Android, so we can't update the version for `file_selector.` Should we remove the `publish_to: none` value from our pubspec and add some `0.0.1` version?

stuartmorgan commented on .github/dependabot.yml:
```diff
@@ -589,3 +589,33 @@ updates:
     open-pull-requests-limit: 10
+
+  - package-ecosystem: "gradle"
+    directory: "/packages/file_selector/file_selector_android/android"
```
Nit: please keep the package entries in this file in alphabetical order; this should go after `espresso`.

stuartmorgan commented on packages/file_selector/file_selector/.metadata:
```diff
@@ -0,0 +1,30 @@
+  revision: 4f9d92fbbdf072a70a70d2179a9f87392b94104c
+  channel: stable
+
+project_type: app
```
This isn't the correct metadata. Please remove this file entirely; we've not been adding this retroactively since it would have misleading data about creation version.

stuartmorgan commented on packages/file_selector/file_selector_android/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+## 0.0.1
```
Nit: we used 0.5.0 for iOS; I think that's a reasonable starting point here as well. 0.0.1 implies that this is extremely experimental/incomplete.

stuartmorgan commented on packages/file_selector/file_selector_android/example/android/app/build.gradle:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    defaultConfig {
+        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
```
Nit: you can remove this and the other template-created informational comments in this file.

stuartmorgan commented on packages/file_selector/file_selector_android/example/lib/get_directory_page.dart:
```diff
@@ -0,0 +1,78 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
I assume the example Dart files are exact copies of the ones in the other implementation packages?

gantelo replied:
Yes! I think they have very few changes but overall a copy. 
We didn´t include the Android version in the ´file_selector´ base package because that would need a release for it to be deployed, plus it would be modifying two packages at once.

We´ll include that in a following PR.

stuartmorgan commented on packages/file_selector/file_selector_android/example/pubspec.yaml:
```diff
@@ -0,0 +1,26 @@
+name: file_selector_android_example
+description: Demonstrates how to use the file_selector_android plugin.
+
+publish_to: 'none' # Remove this line if you wish to publish to pub.dev
```
The comment here can be removed.

stuartmorgan commented on packages/file_selector/file_selector_android/pubspec.yaml:
```diff
@@ -0,0 +1,34 @@
+
+environment:
+  sdk: ">=2.12.0 <3.0.0"
+  flutter: ">=2.10.0"
```
Nit: Let's not support anything older than current `stable` for a new package.

stuartmorgan commented on packages/file_selector/file_selector_android/example/test/widget_test.dart:
```diff
@@ -0,0 +1,66 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
```
This file isn't actually doing anything useful, and should be removed.

stuartmorgan commented on packages/file_selector/file_selector_android/test/file_selector_android_test.dart:
```diff
@@ -0,0 +1,83 @@
+  setUp(() {
+    plugin = FileSelectorAndroid();
+    log = <MethodCall>[];
+    plugin.channel.setMockMethodCallHandler((MethodCall methodCall) async {
```
This will need to be changed for Pigeon usage; see file_selector_windows_test.dart for an example of the preferred structure for new tests of Pigeon code.

stuartmorgan commented on packages/file_selector/file_selector_android/lib/file_selector_android.dart:
```diff
@@ -0,0 +1,148 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/services.dart';
+
+const MethodChannel _channel =
```
This should be using Pigeon, not method channels directly, as we're generally using Pigeon for new code when possible. (You have pigeon in the pubspec, so I'm surprised it's not being used here.)

gantelo replied:
Yes, we explored Pigeon but decided, later on, to keep consistent with the ´image_picker´. We'll modify our approach to use pigeons, shouldn't be too much work 😉.

stuartmorgan replied:
`image_picker` doesn't use Pigeon only because it predates Pigeon, and we haven't devoted time to migrating it yet.

gantelo replied:
We made a refactor to use pigeons, let us know what you think!

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/MethodResultWrapper.java:
```diff
@@ -0,0 +1,54 @@
+import androidx.annotation.NonNull;
+import io.flutter.plugin.common.MethodChannel;
+
+public class MethodResultWrapper implements MethodChannel.Result {
```
This class shouldn't be necessary in Flutter 3+.

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/PathUtils.java:
```diff
@@ -0,0 +1,278 @@
+
+    if (uri.getScheme().equalsIgnoreCase("content")) {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        return copyFileToInternalStorage(uri, context, cacheFolder);
```
Why would selecting a save location involve copying files? If this is actually necessary, it definitely need an explanatory comment.

(Also, if for some reason we need to copy files, it shouldn't be done on the platform thread.)

gantelo replied:
Addressed in the other comment. [Issue created](https://github.com/flutter/flutter/issues/113441) and save path not included in the first version.

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/PathUtils.java:
```diff
@@ -0,0 +1,278 @@
+
+  /**
+   * Returns an absolute path for the file to be saved from the 3 possible content paths to return.
+   * If none of these match, then we will create a copy of the created file in the internal storage
```
Related to the comment below: I'm not sure what "the created file" is referring to here. What is being created?

gantelo replied:
I'll answer both questions here:



> Related to the comment below: I'm not sure what "the created file" is referring to here. What is being created?

This is related to [this issue](https://github.com/flutter/flutter/issues/111583) for iOS. Basically, for mobile, we need to set a save path for a user to save a file. However, we must create an empty file to be able to write in, and then return the file path.

If you think this is a foul functionality, we may not include it in this first version, as iOS did (or didn't), and then discuss what can be done about it.



> Why would selecting a save location involve copying files? If this is actually necessary, it definitely need an explanatory comment.
> 
> (Also, if for some reason we need to copy files, it shouldn't be done on the platform thread.)

First, copying files to the internal storage is:
 - Done in Android Q+
 - Not actual internal storage, but the cache of the app using this package

Android Q+ requires new files or access to files to be accessed from the cache. So this might be a polemic addition to the package.
However, we were thinking of this from a user perspective, where you want (using flutter in Android) to prompt a user to save, for example, an image, so it is an addition that might be useful in the future.

I agree to keep it out of the platform thread; if you think this functionality can be implemented now, we can think of another way. Otherwise, as I mentioned before, we can leave it as an issue to tackle in a post-deployment.

stuartmorgan replied:
> Basically, for mobile, we need to set a save path for a user to save a file. However, we must create an empty file to be able to write in, and then return the file path.
> 
> If you think this is a foul functionality, we may not include it in this first version, as iOS did (or didn't), and then discuss what can be done about it.

We should not include this implementation if it's that fundamentally different from how this works on other platforms. If/when there's actually demand for this functionality on mobile, we can evaluate an API that actually matches the platform semantics.

gantelo replied:
Ok! I'll remove this and leave it as it is in iOS, and we could probably create an issue to track it in the future, what do you think?

stuartmorgan replied:
Making an issue (to capture the investigation done so far, and to provide people something to vote for if they want it) sounds good to me.

gantelo replied:
[Created!](https://github.com/flutter/flutter/issues/113441)

stuartmorgan-g reviewed (changes requested):
Thanks for the PR! Sorry I didn't get a chance to review the earlier structure PR; I haven't been able to keep up with reviews recently. Some specific notes below (many just nits, but I wanted to flag them now to avoid extra rounds later), but some high-level notes:
- It doesn't look like there are any integration tests here. Isn't this testable with native Android integration tests?
- Per some inline comments below, this should use Pigeon.
- This new package will need a CODEOWNERS entry. @GaryQian is probably the best option since there's similarity to `image_picker`.

@GaryQian Could you do a high-level review of the Java code here? The details of the interface with Dart will change for Pigeon, but feedback on the overall implementation would be good to get started on.

gantelo commented:
@stuartmorgan Thanks a lot for the feedback!
I didn't reply to some of the comments to not spam you, but we'll tackle them in a first commit and then work on the heavier stuff.

gantelo commented:
Hi @stuartmorgan, we made several changes since the last review.

Now we're using the Pigeon library and addressed several comments.

Also, we added @GaryQian as a CodeOwner.

Lastly, about the android native integration tests. What would you like to see? So far, we have added unit tests for the plugin and the delegate, asserting the Activities and Intents to be the expected ones.
So, we're unsure what exactly we should be testing there. Since the integration tests in the example would be like re-testing what we already have, does mocking the platform make sense?

Maybe we're not seeing it clearly and need some guidance there.

Regards

stuartmorgan commented:
Why are there so many visible-for-test methods in the code here? It seems like there's a lot of testing that the implementation is structured a specific way rather than testing that the API works as intended, which is a testing anti-pattern.

Compare with the macOS or Windows unit tests, for example, which test the entire native flow and don't use any internals.

> Lastly, about the android native integration tests. What would you like to see?

Ideally, one full end-to-end test of each supported method to ensure that the complete flow works.

> Since the integration tests in the example would be like re-testing what we already have

Unit tests cannot ensure that the Dart->native->Dart round trip actually works correctly. While using Pigeon mitigates potential bugs there, it's still not error-proof.

gantelo commented:
@stuartmorgan Cool!
We're working on the integration tests, using espresso to mock our way into the intents in Android.
I'll ping you once that's solved, but expect it for this iteration.

As for the @VisibleForTesting in the internal functions, we removed the structure testing and left only the method testing.

The tests are testing in the following way:
- For the delegate, which does the "heavy work":
  - Call `onActivityResult` to simulate that the intent succeeded and a result was returned, and then expect the value returned to be the one we mocked.
  - Test these calls for every branch
- For the plugin itself:
  - Test the setup, initialize, and teardown is made correctly
  - Test the intents are called with the correct parameters and the results are called too

gantelo commented:
Hi @stuartmorgan,

We added the remaining integration tests as requested. Thank you for your quick response to our concerns there.

Please take a new look into this when you have some time, or maybe @GaryQian can do an Android review since we added him as a code owner.

Regards

yegor-pelykh commented:
Could you clarify please when support for Android will be ready with the ability to save as well?

stuartmorgan commented:
> with the ability to save as well?

There are currently no plans to add `getSavePath` support on mobile; you are welcome to vote for https://github.com/flutter/flutter/issues/113441 and add a comment there explaining your use case.

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/ActivityStateHelper.java:
```diff
@@ -0,0 +1,87 @@
+import io.flutter.embedding.engine.plugins.lifecycle.FlutterLifecycleAdapter;
+
+/**
+ * Move all activity-lifetime-bound states into this helper object, so that {@code setup} and {@code
```
"Owns all activity-lifetime-bound state, to simplify lifetime state handling for clients."

(The current comment describes a point-in-time refactor—which never even happened in this plugin since it's newer than the change in the code this is copied from—rather than what the class actually does.)

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+    finishWithListSuccess(srcPaths);
+  }
+
+  private <T> boolean setPendingResult(Messages.Result<T> result) {
```
This needs a declaration comment explaining what it does and returns, especially since the return value is non-obvious.

adpinola replied:
I'm rewriting this.

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+
+  private <T> boolean setPendingResult(Messages.Result<T> result) {
+    if (pendingResult != null) {
+      return true;
```
Also, why is "true" the failure case, rather than "false"?

stuartmorgan commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorPlugin.java:
```diff
@@ -0,0 +1,126 @@
+    return activityState;
+  }
+
+  private void setup(
```
`setUp`; setup is a noun, not a verb, and methods should generally have verb names. (The Pigeon naming is a bug in Pigeon.)

stuartmorgan commented on packages/file_selector/file_selector_android/pubspec.yaml:
```diff
@@ -0,0 +1,34 @@
+issue_tracker: https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+label%3A%22p%3A+file_selector%22
+version: 0.5.0
+
+publish_to: 'none' # Remove this line when we make the first publish to pub.dev
```
This should be removed; landing this PR would be when we would want to publish to pub.dev.

stuartmorgan-g reviewed (changes requested):
@GaryQian Ping on the Java review here.

adpinola commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+        return false;
+    }
+
+    return true;
```
According to Interface [PluginRegistry.ActivityResultListener](https://api.flutter.dev/javadoc/io/flutter/plugin/common/PluginRegistry.ActivityResultListener.html), method `onActivityResult` should return `true` if the result has been handled.

stuartmorgan replied:
Oh, I misread the switch; mixing early `return`s and `break`s in a switch is kind of confusing. I would suggest replacing the `break`s with `return true`.

adpinola commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+        handleOpenFileResult(resultCode, data);
+        break;
+      default:
+        return false;
```
Well It is expected to this code won't be reached because there are only two `startActivityForResult` calls, one for each pre-defined request code: `REQUEST_CODE_GET_DIRECTORY_PATH` and `REQUEST_CODE_OPEN_FILE`. Thoughts?

stuartmorgan replied:
If it's supposed to be unreachable, then it should probably throw instead of silently failing.

adpinola commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+    openFileIntent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, isMultipleSelection);
+
+    if (acceptedTypeGroups != null && !acceptedTypeGroups.isEmpty()) {
+      openFileIntent.setType(acceptedTypeGroups.get(0));
```
The allowed types are set in a two-step process. First, we call `setType` where we are using the first element from the list.
Next line `openFileIntent.putExtra(Intent.EXTRA_MIME_TYPES, acceptedTypeGroups.toArray(new String[0]))` is adding the extra allowed types to the Intent.

adpinola commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+    handleOpenFileActionResults(srcPaths);
+  }
+
+  private void handleGetDirectoryPathResult(String path) {
```
agree

adpinola committed:
apply feedback

adpinola committed:
change minSdkVersion to 20

adpinola commented:
Changes applied, @gantelo please resolve conversations. Thanks.

adpinola committed:
replace break with returns

gantelo commented:
@adpinola I left two unresolved conversations where I didn't see a final response or an "outdated" file, should I also mark them as resolved?

adpinola committed:
enable explicitly multidex for API 20 support

remove Robolectric

adpinola commented:
> @adpinola I left two unresolved conversations where I didn't see a final response or an "outdated" file, should I also mark them as resolved?

It's ok, leave those as opened while we wait for an additional review. Thanks!

stuartmorgan commented on packages/file_selector/file_selector_android/test/file_selector_android_test.dart:
```diff
@@ -0,0 +1,96 @@
+      when(mockApi.openFiles(any))
+          .thenAnswer((_) async => <String>['/path/example/final']);
+    });
+    test('simple call works', () async {
```
Nit: please add a blank line between tests and between groups.

stuartmorgan commented on packages/file_selector/file_selector_android/pubspec.yaml:
```diff
@@ -0,0 +1,32 @@
+
+dev_dependencies:
+  build_runner: 2.1.11
+  flutter_lints: ^2.0.0
```
Please remove this; we don't use flutter_lints.

stuartmorgan commented on packages/file_selector/file_selector_android/example/pubspec.yaml:
```diff
@@ -0,0 +1,30 @@
+  flutter_driver:
+    sdk: flutter
+  flutter_plugin_android_lifecycle: ^2.0.1
+  permission_handler: ^10.0.1
```
Why is this needed? We don't generally depend on packages that we don't control.

stuartmorgan-g reviewed (changes requested):
A few minor notes on the non-Java parts, and it look like a few of my comments from before (e.g., missing declaration comments in Java code) haven't been resolved yet, but otherwise the non-Java parts look good to me. @GaryQian will review the Java shortly.

stuartmorgan commented:
Oh, you'll also need to add this package to `script/configs/exclude_integration_android.yaml` as a known exclusion so that CI doesn't fail on post-submit; see script/configs/exclude_integration_ios.yaml for an example. You can go ahead and add the app-facing package to the list at the same time so that we don't forget in the follow-up PR.

(CI is only passing because the `firebase_test_lab` script is being skipped in presubmit for PRs that are from outside the Flutter team.)

adpinola committed:
apply feedback

adpinola commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/FileSelectorDelegate.java:
```diff
@@ -0,0 +1,217 @@
+    this.pendingResult = result;
+  }
+
+  public void clearCache() {
```
Sure thing, I missed this comment before.

adpinola committed:
add missing docs

change public methods to private

adpinola commented:
Hi @gantelo can you resolve the opened conversations? [This one](https://github.com/flutter/plugins/pull/6547/files#r1028455117) too. Thanks.

stuartmorgan commented:
Really adding @GaryQian for review this time; it looks like I pinged last time without actually adding the review bit, sorry about that.

stuartmorgan commented:
@reidbaker Could you find a new reviewer for this PR?

reidbaker commented:
Added myself and will review this after some others I am on.

reidbaker commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/LifeCycleHelper.java:
```diff
@@ -0,0 +1,68 @@
+import androidx.lifecycle.DefaultLifecycleObserver;
+import androidx.lifecycle.LifecycleOwner;
+
+public class LifeCycleHelper
```
A flutter provided Lifecycle helper does not exist yet but if you find yourself needing a class like this in the future checkout https://docs.google.com/document/d/1wXB1zNzYhd2SxCu1_BK3qmNWRhonTB6qdv4erdtBQqo/edit?resourcekey=0-WOBqqOKiO9SARnziBg28pg

reidbaker commented on packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/PathUtils.java:
```diff
@@ -0,0 +1,112 @@
+
+public class PathUtils {
+
+  static String cacheFolder = "file_selector";
```
Is this assumed to be the same cache as the one defined in [FileSelectorDelegate.java](https://github.com/flutter/plugins/pull/6547/files#diff-233f4faa39ffe628203060d7feae95b567bf35f46d0685fd91dc550e3bc52c56) if so then consider using a shared static variable.

reidbaker reviewed (changes requested):
There are a handful of unused images as well. Can you remove them?

stuartmorgan-g commented:
We've just completed the migration of the plugin code to the flutter/packages repository, as described in https://flutter.dev/go/flutter-plugins-repo-migration, and this repository is now being archived. Unfortunately that means that all in-progress PRs here must be moved to flutter/packages.

Please see [our instructions](https://github.com/flutter/flutter/wiki/Migrating-Plugins-repository-PRs-to-Packages) for an explanation of how to move your PR, and if you have any issues moving your PR please don't hesitate to reach out in the #hackers-ecosystem channel in [Discord](https://github.com/flutter/flutter/wiki/Chat).

Our apologies that your PR was caught in this one-time transition. We're aware that it's disruptive in the short term, and appreciate your help in getting us to a better long-term state!
